stl/inc/spanstream

```diff
@@ -0,0 +1,386 @@
+#endif // __cpp_lib_concepts
+
+    // N4892 [ispanstream.assign], assignment and swap
+    basic_ispanstream& operator=(const basic_ispanstream&) = delete;
+
+    basic_ispanstream& operator=(basic_ispanstream&& _Right) {
+        _Mybase::swap(_Right);
+        _STD swap(_Buf, _Right._Buf);
+    }
+
+    void swap(basic_ispanstream& _Right) {
+        _Mybase::swap(_Right);
+        _STD swap(_Buf, _Right._Buf);
+    }
+
+    // N4892 [ispanstream.members], member functions
+    _Mysb* rdbuf() const noexcept {
+        return const_cast<_Mysb*>(_STD addressof(_Buf));
+    }
+
+    span<const _Elem> span() const noexcept {
+        return this->rdbuf()->span();
+    }
+
+    void span(_STD span<_Elem> _Span) noexcept {
+        this->rdbuf()->span(_Span);
+    }
+
+#ifdef __cpp_lib_concepts
+    // clang-format off
+    template <_RANGES borrowed_range _ReadOnlyRange>
+        requires (!convertible_to<_ReadOnlyRange, _STD span<_Elem>>
+                && convertible_to<_ReadOnlyRange, _STD span<const _Elem>>)
+    void span(_ReadOnlyRange&& _Range) noexcept {
+        this->span(
+            _STD span<_Elem>{const_cast<_Elem*>(_RANGES data(_Range)), static_cast<size_t>(_RANGES size(_Range))});
+    }
+    // clang-format on
+#endif // __cpp_lib_concepts
+
+private:
+    _Mysb _Buf;
+};
+
+template <class _Elem, class _Traits>
+void swap(basic_ispanstream<_Elem, _Traits>& _Left, basic_ispanstream<_Elem, _Traits>& _Right) {
+    _Left.swap(_Right);
+}
+
+template <class _Elem, class _Traits>
+class basic_ospanstream : public basic_ostream<_Elem, _Traits> {
+public:
+    using _Mybase     = basic_ostream<_Elem, _Traits>;
+    using _Mysb       = basic_spanbuf<_Elem, _Traits>;
+    using char_type   = _Elem;
+    using int_type    = typename _Traits::int_type;
+    using pos_type    = typename _Traits::pos_type;
+    using off_type    = typename _Traits::off_type;
+    using traits_type = _Traits;
+
+    // N4892 [ospanstream.ctor], constructors
+    explicit basic_ospanstream(_STD span<_Elem> _Span, ios_base::openmode _Which = ios_base::out)
+        : _Mybase(_STD addressof(_Buf)), _Buf(_Span, _Which | ios_base::out) {}
+
+    basic_ospanstream(const basic_ospanstream&) = delete;
+
+    basic_ospanstream(basic_ospanstream&& _Right) : _Mybase(_STD move(_Right)), _Buf(_STD move(_Right._Buf)) {
+        _Mybase::set_rdbuf(_STD addressof(_Buf));
+    }
+
+    // N4892 [ospanstream.assign], assignment and swap
+    basic_ospanstream& operator=(const basic_ospanstream&) = delete;
+
+    basic_ospanstream& operator=(basic_ospanstream&& _Right) {
+        _Mybase::swap(_Right);
+        _STD swap(_Buf, _Right._Buf);
+    }
+
+    void swap(basic_ospanstream& _Right) {
+        _Mybase::swap(_Right);
+        _STD swap(_Buf, _Right._Buf);
+    }
+
+    // N4892 [ospanstream.members], member functions
+    _Mysb* rdbuf() const noexcept {
+        return const_cast<_Mysb*>(_STD addressof(_Buf));
+    }
+
+    span<const _Elem> span() const noexcept {

```

Comment:
Unlike `basic_ispanstream` above, `basic_ospanstream` needs to return `span<_Elem>` here.

Occurs below for `basic_spanstream`.